src/xdsl/dialects/builtin.py

```diff
@@ -190,21 +192,95 @@ def verify(self, data: Data) -> None:
 
 
 @irdl_attr_definition
-class VectorAttr(ParametrizedAttribute):
+class VectorType(ParametrizedAttribute):
     name = "vector"
+
+    shape = ParameterDef(ArrayOfConstraint(IntegerAttr))
+    element_type = ParameterDef(AnyAttr())
+
+    def get_num_dims(self) -> int:
+        return len(self.parameters[0].data)
+
+    def get_shape(self) -> List[int]:
+        return [i.parameters[0].data for i in self.shape.data]
+
+    @staticmethod
+    @builder
+    def from_type_and_list(
+            referenced_type: Attribute,
+            shape: List[Union[int, IntegerAttr]] = None) -> VectorType:
+        if shape is None:
+            shape = [1]
+        return VectorType([
+            ArrayAttr.from_list([IntegerAttr.build(d) for d in shape]),
+            referenced_type
+        ])
+
+    @staticmethod
+    @builder
+    def from_params(
+        referenced_type: Attribute,
+        shape: ArrayAttr = ArrayAttr.from_list(
+            [IntegerAttr.from_int_and_width(1, 64)])
+    ) -> VectorType:
+        return VectorType([shape, referenced_type])
+
+
+@irdl_attr_definition
+class TensorType(ParametrizedAttribute):
+    name = "tensor"
+
+    shape = ParameterDef(ArrayOfConstraint(IntegerAttr))
+    element_type = ParameterDef(AnyAttr())
+
+    def get_num_dims(self) -> int:
+        return len(self.parameters[0].data)
+
+    def get_shape(self) -> List[int]:
+        return [i.parameters[0].data for i in self.shape.data]
+
+    @staticmethod
+    @builder
+    def from_type_and_list(
+            referenced_type: Attribute,
+            shape: List[Union[int, IntegerAttr]] = None) -> TensorType:
+        if shape is None:
+            shape = [1]
+        return TensorType([
+            ArrayAttr.from_list([IntegerAttr.build(d) for d in shape]),
+            referenced_type
+        ])
+
+    @staticmethod
+    @builder
+    def from_params(
+        referenced_type: Attribute,
+        shape: ArrayAttr = ArrayAttr.from_list(
+            [IntegerAttr.from_int_and_width(1, 64)])
+    ) -> TensorType:
+        return TensorType([shape, referenced_type])
+
+
+@irdl_attr_definition
+class DenseIntOrFPElementsAttr(ParametrizedAttribute):
```

Comment:
Can we add builders for vector and tensor? For example, to create a new tensor containing x integers, it would be nice to not have to build separately the `TensorType`.

Reply:
Extracting the shape from the passed list would be enough, I assume.

Reply:
How should one specify a variable dimensional vector with python type annotations?